### PR TITLE
Do not allow dropping Extended Master Secret extension on renegotiaton

### PR DESCRIPTION
**Pull request by t8m**
I've verified with tlsfuzzer that the handshake failure alert is properly sent.

##### Checklist

- [x] tests are added or updated

### CODE REVIEW

**Comment by t8m** [comment]
The test passes. What I am curious about is that I see:
```
            # ERROR: (int) 'result->server_alert_sent == result->server_alert_received' failed @ test/ssl_test.c:68
            # [552] compared to [0]
            # INFO:  @ test/ssl_test.c:70
            # Server sent alert handshake failure but client received no alert.
```
In the verbose test output. I wonder why. But the tlsfuzzer test receives the handshake failure alert so I suppose this is just some glitch in the test.

File: ssl/statem/extensions.c [hunk 1]
```diff
@@ -1169,14 +1169,29 @@ static int init_etm(SSL *s, unsigned int context)
 
 static int init_ems(SSL *s, unsigned int context)
 {
-    s->s3.flags &= ~TLS1_FLAGS_RECEIVED_EXTMS;
+    if (s->s3.flags & TLS1_FLAGS_RECEIVED_EXTMS) {
+        s->s3.flags &= ~TLS1_FLAGS_RECEIVED_EXTMS;
+        if (s->server)
```
[the remark this answers is not in the document]
Comment (t8m) [reply ssl/statem/extensions.c:1174]: Yeah, I was thinking about it as well. It makes sense to disallow on client as well.

**Comment by t8m** [comment]
> Will there be a separate backport PR to bring 1.1.1 in line?

Yes, once this one is approved.
 
> Should there be a CHANGES entry?

I am not sure it is necessary as it is just about handling a true corner case. No normal client/server would do this.

**Comment by mattcaswell** [comment]
> But I can add Changes entry if it is preferred.

I'm mainly concerned about the 1.1.1 case. It is a change in behaviour and there are weird configurations with multiple servers in a server farm that might behave unexpectedly.

**Comment by t8m** [comment]
> > But I can add Changes entry if it is preferred.
> 
> I'm mainly concerned about the 1.1.1 case. It is a change in behaviour and there are weird configurations with multiple servers in a server farm that might behave unexpectedly.

That would be perhaps relevant for session resumptions (but the mismatch of ExtMS support on session resumption is already disallowed). But I do not think you can switch a server on a single TLS connection during the renegotiation.

File: ssl/statem/extensions.c (hunk 1 above)
[the remark this answers is not in the document]
Comment (t8m) [reply ssl/statem/extensions.c]: Applied on client side as well.

**Comment by t8m** [comment]
Anyway, I've added the CHANGES entry, just in case.
Unfortunately it is impossible to test the client side check for now as the server side checks do not allow to drop the EMS on renegotiation from the server side. But I've manually verified it works by hacking out the checks.
@mattcaswell final review or any other comments?

**Comment by davidben** [comment]
We also forbid *adding* EMS on renegotiation if the initial handshake did not negotiate it. That way there aren't weird problems if, say, the caller queries EMS existence after renego and assumes some property of the earlier data.

**Comment by t8m** [comment]
@mattcaswell ping?

**Comment by t8m** [comment]
Merged to master. PR for 1.1.1 coming.

**Pull request closed without merging**